upstairs/src/volume.rs

```diff
@@ -1238,25 +1246,167 @@ impl Volume {
+                            "The VCRs have no meaningful difference"
+                        )
+                    }
+                    VCRDelta::Generation => Ok(None),
+                }
+            }
+            VCRDelta::Generation => {
+                // The sub_volume VCR is the same as what we have, but with
+                // updated generation numbers.  Check what the ROP found.
+                match rop_result {
+                    VCRDelta::Target { old, new } => {
+                        // Just the ROP has changes, so let's return those.
+                        Ok(Some((old, new)))
+                    }
+                    _ => Ok(None),
+                }
+            }
+            VCRDelta::NewMissing => {
+                // This should not be possible for sub volumes as the VCR
+                // is a required argument to this function.
+                panic!("New VCR is missing when comparing regions to replace");
+            }
+        }
+    }
+
+    // Given two VCRs where we expect the only type to be a
+    // VolumeConstructionRequest::Region.  The VCR can be from a sub_volume
+    // or a read_only_parent.  The caller is expected to know how to
+    // handle the result depending on what it sends us.
+    //
+    // We return:
+    // VCRDelta::Same
+    // If the two VCRs are identical
+    //
+    // VCRDelta::Generation
+    // If it's just a generation number increase in the new VCR.
+    //
+    // VCRDelta::NewMissing
+    // If the new VCR is missing (None).
+    //
+    // VCRDelta::Target(old_target, new_target)
+    // If we have both a generation number increase, and one and only
+    // one target is different in the new VCR.
+    //
+    // Any other difference is an error, and an error returned here means the
+    // VCRs are incompatible in a way that prevents one from replacing another.
+    fn compare_vcr_region_for_replacement(
+        opt_o_vol: Option<Box<VolumeConstructionRequest>>,
+        opt_n_vol: Option<Box<VolumeConstructionRequest>>,
+    ) -> Result<VCRDelta, CrucibleError> {
+        if opt_n_vol.is_some() {
+            if opt_o_vol.is_none() {
+                // We support a VCR where a read_only_parent has gone away,
+                // but we should never see addition of a read_only_parent
+                // to a VCR where there was not one originally.
+                crucible_bail!(
+                    ReplaceRequestInvalid,
+                    "VCR added where there should not be one"
+                )
+            }
+        } else {
+            // New VCR is none. When comparing read_only_parents this is
+            // okay.  We indicate that with this return status and let the
+            // caller decide what action to take
+            return Ok(VCRDelta::NewMissing);
+        }
+
+        // To get here, both VCRs are populated.
+        let o_vol = opt_o_vol.unwrap();
+        let n_vol = opt_n_vol.unwrap();
```

Comment:
I still prefer a tuple match here, which is also shorter:
```suggestion
        let (o_vol, n_vol) = match (opt_o_vol, opt_n_vol) {
            (.., None) => {
                // New VCR is none. When comparing read_only_parents this is
                // okay.  We indicate that with this return status and let the
                // caller decide what action to take
                return Ok(VCRDelta::NewMissing);
            }
            (None, Some(..)) => 
                // We support a VCR where a read_only_parent has gone away,
                // but we should never see addition of a read_only_parent
                // to a VCR where there was not one originally.
                crucible_bail!(
                    ReplaceRequestInvalid,
                    "VCR added where there should not be one"
                )
            }
            (Some(o_vol), Some(n_vol)) => (o_vol, n_vol),
        };
```

Reply:
Done in latest commit